File: vibeshield-ts/src/core/jwt-validator.ts
/**
 * VibeShield JWT Secret Key Validator
 *
 * Detects weak, hardcoded, or insecure JWT secret keys that AI code generators
 * commonly produce. This module performs validation only — it does NOT encode
 * or decode JWT tokens.
 *
 * Zero external dependencies.
 */

/** Result of a JWT secret validation check. */
export interface JwtValidationResult {
  /** Whether the secret passes all mandatory checks. */
  valid: boolean;
  /** Critical issues that must be fixed (blocks usage). */
  errors: string[];
  /** Non-critical observations the developer should review. */
  warnings: string[];
}

/**
 * Well-known weak secrets that AI models frequently hardcode.
 * All comparisons are case-insensitive.
 */
const HARDCODED_PATTERNS: ReadonlyArray<string> = [
  'secret',
  'password',
  '123456',
  '1234567890',
  'qwerty',
  'admin',
  'root',
  'jwt_secret',
  'jwt_key',
  'token_secret',
  'auth_secret',
  'mysecret',
  'my_secret',
  'my-secret',
  'supersecret',
  'super_secret',
  'changeme',
  'change_me',
  'test',
  'testing',
  'default',
  'example',
  'placeholder',
  'your_secret_here',
  'your-secret-here',
  'replace_me',
  'todo',
  'fixme',
  'key',
  'apikey',
  'api_key',
];

/** Minimum acceptable secret length. */
const MIN_SECRET_LENGTH = 10;

/** Minimum recommended secret length for strong security. */
const RECOMMENDED_SECRET_LENGTH = 32;

/**
 * Calculates Shannon entropy (bits per character) for a given string.
 * Higher entropy indicates more randomness/complexity.
 */
function calculateShannonEntropy(secret: string): number {
  if (!secret) return 0;
  
  const frequencyMap = new Map<string, number>();
  for (const char of secret) {
    frequencyMap.set(char, (frequencyMap.get(char) || 0) + 1);
  }
  
  let entropy = 0;
  const len = secret.length;
  
  for (const count of frequencyMap.values()) {
    const probability = count / len;
    entropy -= probability * Math.log2(probability);
  }
  
  return entropy;
}


/**
 * Validates a JWT secret key for common security vulnerabilities.
 *
 * This function checks for:
 * - Missing or empty secrets
 * - Known hardcoded/placeholder values
 * - Insufficient length
 * - Low character diversity (repeating chars, all-numeric, all-lowercase)
 * - Low Shannon entropy
 *
 * @param secret - The JWT secret string to validate (may be undefined).
 * @returns A {@link JwtValidationResult} with errors and warnings.
 */
export function validateJwtSecret(secret: string | undefined): JwtValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  // ── 1. Undefined / Empty Check ──────────────────────────────────────
  if (secret === undefined || secret === null) {
    errors.push('JWT secret is undefined. Provide a secret via environment variable.');
    return { valid: false, errors, warnings };
  }

  const trimmed = secret.trim();
  if (trimmed.length === 0) {
    errors.push('JWT secret is empty or contains only whitespace.');
    return { valid: false, errors, warnings };
  }

  const withoutPadding = trimmed.replace(/=+$/, '');

  // ── 2. Hardcoded Pattern Detection ──────────────────────────────────
  const lower = trimmed.toLowerCase();
  for (const pattern of HARDCODED_PATTERNS) {
    if (lower === pattern) {
      errors.push(`JWT secret matches a well-known weak value: "${pattern}". Use a cryptographically random secret.`);
      break;
    }
  }

  // ── 3. Length Check ─────────────────────────────────────────────────
  if (trimmed.length < MIN_SECRET_LENGTH) {
    errors.push(`JWT secret is too short (${trimmed.length} chars). Minimum required: ${MIN_SECRET_LENGTH} characters.`);
  } else if (trimmed.length < RECOMMENDED_SECRET_LENGTH) {
    warnings.push(`JWT secret is ${trimmed.length} chars. Recommended: ${RECOMMENDED_SECRET_LENGTH}+ characters for production use.`);
  }

  // ── 4. Repeating Character / Pattern Check & Sequential Patterns ───
  // Repetitive characters (aaa, 111)
  if (/(.)\1{2,}/.test(withoutPadding)) {
    errors.push('Contains repetitive characters');
  }

  // Repeated patterns (abcabc, 123123)
  if (/(.+)\1+/.test(withoutPadding)) {
    errors.push('Contains repeated patterns');
  }

  // Sequential patterns (abc, 123, qwerty)
  const sequentialPatterns = [
    /abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz/i,
    /012|123|234|345|456|567|678|789|890/,
    /qwerty|asdfgh|zxcvbn|qazwsx/i,
  ];

  for (const pattern of sequentialPatterns) {
    if (pattern.test(withoutPadding)) {
      errors.push('Contains sequential patterns');
      break;
    }
  }

  // ── 5. Character Class Diversity ────────────────────────────────────
  const hasUpper = /[A-Z]/.test(trimmed);
  const hasLower = /[a-z]/.test(trimmed);
  const hasDigit = /\d/.test(trimmed);
  const hasSpecial = /[^A-Za-z0-9]/.test(trimmed);

  // All-numeric
  if (/^\d+$/.test(trimmed)) {
    errors.push('JWT secret contains only digits. Use a mix of letters, numbers, and symbols.');
  }

  // All same-case letters
  if (/^[a-z]+$/.test(trimmed) && trimmed.length >= MIN_SECRET_LENGTH) {
    warnings.push('JWT secret contains only lowercase letters. Consider adding uppercase, digits, and symbols.');
  }
  if (/^[A-Z]+$/.test(trimmed) && trimmed.length >= MIN_SECRET_LENGTH) {
    warnings.push('JWT secret contains only uppercase letters. Consider adding lowercase, digits, and symbols.');
  }

  // ── 6. Shannon Entropy Check ────────────────────────────────────────
  const entropy = calculateShannonEntropy(withoutPadding);
  if (entropy < 3.5) {
    errors.push('Low entropy detected');
    warnings.push(`Low entropy detected (Shannon entropy: ${entropy.toFixed(2)})`);
  }

  // ── 7. Complexity recommendation ────────────────────────────────────
  const classCount = [hasUpper, hasLower, hasDigit, hasSpecial].filter(Boolean).length;
  if (classCount < 3 && trimmed.length >= MIN_SECRET_LENGTH && errors.length === 0) {
    warnings.push('JWT secret uses fewer than 3 character classes. Mix uppercase, lowercase, digits, and symbols.');
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}

File: vibeshield-python/src/jwt_validator.py
"""
VibeShield JWT Secret Key Validator (Python)

Detects weak, hardcoded, or insecure JWT secret keys that AI code generators
commonly produce. This module performs validation only — it does NOT encode
or decode JWT tokens.

Zero external dependencies.
"""

import math
import re
from typing import Optional

# Well-known weak secrets that AI models frequently hardcode.
# All comparisons are case-insensitive.
HARDCODED_PATTERNS: list[str] = [
    "secret",
    "password",
    "123456",
    "1234567890",
    "qwerty",
    "admin",
    "root",
    "jwt_secret",
    "jwt_key",
    "token_secret",
    "auth_secret",
    "mysecret",
    "my_secret",
    "my-secret",
    "supersecret",
    "super_secret",
    "changeme",
    "change_me",
    "test",
    "testing",
    "default",
    "example",
    "placeholder",
    "your_secret_here",
    "your-secret-here",
    "replace_me",
    "todo",
    "fixme",
    "key",
    "apikey",
    "api_key",
]

# Minimum acceptable secret length.
MIN_SECRET_LENGTH = 10

# Minimum recommended secret length for strong security.
RECOMMENDED_SECRET_LENGTH = 32


def calculate_shannon_entropy(secret: str) -> float:
    """Calculates Shannon entropy (bits per character) for a given string."""
    if not secret:
        return 0.0
    
    frequency: dict[str, int] = {}
    for char in secret:
        frequency[char] = frequency.get(char, 0) + 1
    
    entropy = 0.0
    length = len(secret)
    
    for count in frequency.values():
        probability = count / length
        entropy -= probability * math.log2(probability)
    
    return entropy



def validate_jwt_secret(secret: Optional[str]) -> dict:
    """
    Validates a JWT secret key for common security vulnerabilities.

    Checks for:
    - Missing or empty secrets
    - Known hardcoded/placeholder values
    - Insufficient length
    - Low character diversity (repeating chars, all-numeric, all-lowercase)
    - Low Shannon entropy

    Args:
        secret: The JWT secret string to validate (may be None).

    Returns:
        dict with keys: "valid" (bool), "errors" (list[str]), "warnings" (list[str])
    """
    errors: list[str] = []
    warnings: list[str] = []

    # ── 1. None / Empty Check ──────────────────────────────────────────
    if secret is None:
        errors.append("JWT secret is undefined. Provide a secret via environment variable.")
        return {"valid": False, "errors": errors, "warnings": warnings}

    trimmed = secret.strip()
    if len(trimmed) == 0:
        errors.append("JWT secret is empty or contains only whitespace.")
        return {"valid": False, "errors": errors, "warnings": warnings}

    without_padding = trimmed.rstrip('=')

    # ── 2. Hardcoded Pattern Detection ─────────────────────────────────
    lower = trimmed.lower()
    for pattern in HARDCODED_PATTERNS:
        if lower == pattern:
            errors.append(
                f'JWT secret matches a well-known weak value: "{pattern}". '
                f"Use a cryptographically random secret."
            )
            break

    # ── 3. Length Check ────────────────────────────────────────────────
    if len(trimmed) < MIN_SECRET_LENGTH:
        errors.append(
            f"JWT secret is too short ({len(trimmed)} chars). "
            f"Minimum required: {MIN_SECRET_LENGTH} characters."
        )
    elif len(trimmed) < RECOMMENDED_SECRET_LENGTH:
        warnings.append(
            f"JWT secret is {len(trimmed)} chars. "
            f"Recommended: {RECOMMENDED_SECRET_LENGTH}+ characters for production use."
        )

    # ── 4. Repeating Character / Pattern Check & Sequential Patterns ───
    # Repetitive characters (aaa, 111)
    if re.search(r'(.)\1{2,}', without_padding):
        errors.append('Contains repetitive characters')

    # Repeated patterns (abcabc, 123123)
    if re.search(r'(.+)\1+', without_padding):
        errors.append('Contains repeated patterns')

    # Sequential patterns (abc, 123, qwerty)
    sequential_patterns = [
        r'abc|bcd|cde|def|efg|fgh|ghi|hij|ijk|jkl|klm|lmn|mno|nop|opq|pqr|qrs|rst|stu|tuv|uvw|vwx|wxy|xyz',
        r'012|123|234|345|456|567|678|789|890',
        r'qwerty|asdfgh|zxcvbn|qazwsx',
    ]

    for pattern in sequential_patterns:
        if re.search(pattern, without_padding, re.IGNORECASE):
            errors.append('Contains sequential patterns')
            break

    # ── 5. Character Class Diversity ───────────────────────────────────
    has_upper = bool(re.search(r"[A-Z]", trimmed))
    has_lower = bool(re.search(r"[a-z]", trimmed))
    has_digit = bool(re.search(r"\d", trimmed))
    has_special = bool(re.search(r"[^A-Za-z0-9]", trimmed))

    # All-numeric
    if re.fullmatch(r"\d+", trimmed):
        errors.append("JWT secret contains only digits. Use a mix of letters, numbers, and symbols.")

    # All same-case letters
    if re.fullmatch(r"[a-z]+", trimmed) and len(trimmed) >= MIN_SECRET_LENGTH:
        warnings.append(
            "JWT secret contains only lowercase letters. "
            "Consider adding uppercase, digits, and symbols."
        )
    if re.fullmatch(r"[A-Z]+", trimmed) and len(trimmed) >= MIN_SECRET_LENGTH:
        warnings.append(
            "JWT secret contains only uppercase letters. "
            "Consider adding lowercase, digits, and symbols."
        )

    # ── 6. Shannon Entropy Check ───────────────────────────────────────
    entropy = calculate_shannon_entropy(without_padding)
    if entropy < 3.5:
        errors.append("Low entropy detected")
        warnings.append(f"Low entropy detected (Shannon entropy: {entropy:.2f})")

    # ── 7. Complexity recommendation ───────────────────────────────────
    class_count = sum([has_upper, has_lower, has_digit, has_special])
    if class_count < 3 and len(trimmed) >= MIN_SECRET_LENGTH and len(errors) == 0:
        warnings.append(
            "JWT secret uses fewer than 3 character classes. "
            "Mix uppercase, lowercase, digits, and symbols."
        )

    return {
        "valid": len(errors) == 0,
        "errors": errors,
        "warnings": warnings,
    }
